Classes/URKArchive.mm

```diff
@@ -331,7 +371,61 @@ - (nullable NSURL *)firstVolumeURL {
 
 - (nullable NSArray<NSString*> *)listVolumePaths:(NSError **)error
 {
-    return @[];
+    NSMutableSet<NSString*> *volumePaths = [NSMutableSet new];
+    
+    NSURL * firstVolumeURL = [self firstVolumeURL];
+    if (firstVolumeURL != self.fileURL)
+    {
+        @try {
+            NSError *firstVolumeError = nil;
+            
+            if ([self _unrarOpenFile:firstVolumeURL.path
```

Comment:
This is only a little bit better than calling `[self init...]`. Calling this method changes private members, that now correspond to a different archive's file. How about instead of using `self` at all, make a new `URKArchive` instance here using `firstVolumeURL`? You can still call private methods on it. Perhaps make another private method named `searchForVolumePaths` (or something like that) that returns the list of volume paths. Use `-listFileInfo:` as an example. You can just return the result of that method on the `firstVolumeArchive` instance instead of this entire `@try` block.

Another approach could be to only ever use the first volume in `init...`. If a user initializes an archive with an extension that contains a number, replace the filepath with a path to the first volume. You can refactor the guts of `firstVolumePath` to be a class method that takes a path as an argument, and call that from the instance method with `self.filename` as its argument. You would do those checks before any member variables get assigned.

What do you think?

Reply:
Please take a look at this suggestion. It's the only real thing holding this back. Thanks for hopping back on this request!

Reply:
It looks like you actually did reply to this (I'm seeing it in an email, but not on the page).

> I like a loot this idea. But maybe a init overload like - (instancetype)initWithPath:(NSString *)filePath checkFirstVolume:(bool)checkFirstVolume error:(NSError **)error so the developer can decide wether to use this feature?

I can think of two reasons to make it a separate overload:

1. The old behavior is desirable for some use cases
2. In case the new version has a bug

Regarding 1, I can't think of any reason to ever operate on a subsequent volume instead of the original. I think this is how most RAR extractors work. Can you think of any?

2 may possibly be a concern, but we should make sure the updated `init` gracefully falls back to the existing behavior, logging as it goes to help a developer figure out what may have happened. Basically, do and log each of the following:

* Look for a new root volume if it looks like a subsequent part of a multivolume archive
* If yes to the above, look for the root archive
* If not found, continue as if the one given is the root
* If found, substitute the given path for the root volume path

Thoughts?

Reply:
I did replied... but then messed with the reply...

1. Thought about volume reconstruction, in case the first part does not exist. But with the code as is right now, it will not return a first part if it does not exist, so no case that I can think here.
2. Maybe just adding the correct logs on the first volume check will do, as you say.